File: web/tests/reference/tries/design-add-and-search-words.py
class WordDictionary:
    def __init__(self):
        self.root = {}

    def add_word(self, word):
        node = self.root
        for ch in word:
            node = node.setdefault(ch, {})
        node["$"] = True

    def search(self, word):
        def dfs(i, node):
            if i == len(word):
                return "$" in node
            ch = word[i]
            if ch == ".":
                for key, child in node.items():
                    if key != "$" and dfs(i + 1, child):
                        return True
                return False
            return ch in node and dfs(i + 1, node[ch])

        return dfs(0, self.root)

File: web/tests/reference/tries/design-add-and-search-words.js
class WordDictionary {
  constructor() {
    this.root = new Map();
  }

  addWord(word) {
    let node = this.root;
    for (const ch of word) {
      if (!node.has(ch)) node.set(ch, new Map());
      node = node.get(ch);
    }
    node.set("$", true);
  }

  search(word) {
    const dfs = (i, node) => {
      if (i === word.length) return node.has("$");
      const ch = word[i];
      if (ch === ".") {
        for (const [key, child] of node) {
          if (key !== "$" && dfs(i + 1, child)) return true;
        }
        return false;
      }
      return node.has(ch) && dfs(i + 1, node.get(ch));
    };
    return dfs(0, this.root);
  }
}
